### PR TITLE
Kill conflicting borrows of places with projections.

### LINKED ISSUE #62007: NLL iterating and updating `&mut` fails for non-local Place.
Spawned off of discussion with @ecstatic-morse  and their work on PR #61787 

There is a dataflow subroutine for `Borrows`, `kill_borrows_on_place`, added in PR #56649 to resolve #46589. `kill_borrows_on_place` reads from the `on_entry` field of `BlockSets` (even though I am pretty sure no one was ever supposed to do so; its inclusion in this part of the API was a wart):

https://github.com/rust-lang/rust/blob/f0c2bdf52e0c8bce258f1bbb5652c9691b7f3193/src/librustc_mir/dataflow/impls/borrows.rs#L205-L214

I think this was written with the assumption that `sets.on_entry.clone().union(sets.gen_set)` (aka *on-entry* U *gen-set*) would represent an upper-bound on the things that should be in the kill set. (And then the subsequent loop determines which of those things to actually *put into* the kill set.)

The problem with that is that when we are computing the transfer-function initially, `on_entry` is going to be the empty-set, at least for almost all basic blocks. That all you have to work with in dataflow. So you end up with a kill-set here that is an under-approximation to what it should actually be, if you are assuming that we're supposed to start with the complete (post dataflow) on-entry and combine that with gen-set to get the upper bound on what to use for the kill-set.

Why hasn't this been a problem for us so far? Well, here's my theory:

 1. With respect to soundness: if the kill-set for `Borrows` is a subset of what it should "truly be", then I think the compiler ends up treating borrows as living longer than they should, and so it ends up detecting conflicts that are not actually there, and so the compiler rejects code that you might have other expected it to accept.

 2. With respect to expressiveness: We don't actually get into the problematic part of `kill_borrows_on_place` as often as you might think, because there's a special dodge at the start of the function:

https://github.com/rust-lang/rust/blob/f0c2bdf52e0c8bce258f1bbb5652c9691b7f3193/src/librustc_mir/dataflow/impls/borrows.rs#L196-L203

----

You combine the two points made above, and you end up with things like the following example: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=3d29327853729ec280799760d5e58003

<details>
<summary> Click to see the example Rust code </summary>

```rust
struct List<T> {
    value: T,
    next: Option<Box<List<T>>>,
}


fn to_refs1<T>(mut list: &mut List<T>) -> Vec<&mut T> {
    let mut result = vec![];
    loop {
        result.push(&mut list.value);
        if let Some(n) = list.next.as_mut() {
            list = n;
        } else {
            return result;
        }
    }
}

fn to_refs2<T>(mut list: (&mut List<T>,)) -> Vec<&mut T> {
    let mut result = vec![];
    loop {
        result.push(&mut (list.0).value);
        if let Some(n) = (list.0).next.as_mut() {
            list.0 = n;
        } else {
            return result;
        }
    }
}
```

</details>

(In that example, `to_refs2` is basically the same as `to_refs1`, except that its wrapped in a unary tuple `(T,)`. I think the intention is that if `to_refs1` is accepted, then so should `to_refs2`. But NLL currently accepts `to_refs1` and )

### PR DESCRIPTION
**Pull request by ecstatic-morse**
Resolves #62007.

Due to a bug, the previous version of this check did not actually kill all conflicting borrows unless the borrowed place had no projections. Specifically, `sets.on_entry` will always be empty when `statement_effect` is called. It does not contain the set of borrows which are live at this point in the program.

@pnkfelix describes why this was not caught before in #62007, and created an example where the current borrow checker failed unnecessarily. This PR adds their example as a test, but they will likely want to add some additional ones.

r? @pnkfelix

### CODE REVIEW

File: src/test/run-pass/borrowck/borrowck-borrow-of-projection-kills-other-borrows-issue-62007.rs [hunk 1]
```diff
@@ -0,0 +1,21 @@
+// run-pass
```
Comment (Centril) [inline src/test/run-pass/borrowck/borrowck-borrow-of-projection-kills-other-borrows-issue-62007.rs]: A comment would be good :)

**Comment by ecstatic-morse** [comment]
This still needs some docs in the test case pointing to #46589, as well as a comment explaining why it's sound not to kill any borrows, but I'll wait for @pnkfelix to weigh in here.

File: src/test/run-pass/borrowck/borrowck-borrow-of-projection-kills-other-borrows-issue-62007.rs (hunk 1 above)
Comment (pnkfelix) [reply src/test/run-pass/borrowck/borrowck-borrow-of-projection-kills-other-borrows-issue-62007.rs]: (to clarify: it tends to be a good idea to not rely solely on the filename for a test for the description of what its testing. You can just transcribe part of the description from #62007 to explain it.)

**Comment by pnkfelix** [comment]
The test passed, so that's good.

It would be good to add tests of each of the other kinds of projections, and check how they are handled.

The test you've added here handles `ProjectionElem::Field`.

At the very least, I'd like to see:
 * `Deref` by passing in `Box<&mut List<T>>` instead of `(&mut List<T>,)` (and adjusting the test by adding `*value` where necessary). I expect this one to behave just like the tuple.
 * `ConstantIndex` by passing in `[&mut List<T>; 2]` (and adjusting the test by adding `value[0]` where necessary). I expect this one to behave just like the tuple.
 * `Index` by passing in `[&mut List<T]>; 2]` and adjusting the test by adding `value[i]` where `i` is a local variable. I am not expecting this one to be accepted; it will probably need to be a compile-fail test.
 
It would of course be good to cover the other `ProjectionElem` variants: `Downcast`, `Subslice`; but they would require a little bit more adjustment to the test I think, so I cannot dash them off at the top of my head.

**Comment by pnkfelix** [comment]
(the code changes here look good, too. I'll actually see about adding the tests I described so that we can get this into the pipeline to land sooner rather than later.)

**Comment by pnkfelix** [comment]
I tried to make a variant of the tests I added above that covered the case of a downcast-projection, but I'm not sure if I know a way to get a downcast projection onto the righthand side of an assignment in the form that we need here to observe the phenomenon being tested.

So I'm not going to worry about adding a test for that case at this point.

(LIkewise, I'm not going to worry about subslice projection either.)

**Comment by pnkfelix** [comment]
r=me assuming the tests I added all pass the CI.

**Comment by pnkfelix** [comment]
@bors rollup=never

(This PR may affect performance. I don't think it will, so I'm not blocking it on a perf run. But I also want it to be apparent in the graphs if this PR *does* affect performance.)

**Comment by pnkfelix** [comment]
@centril did I mistype the command for indicating this PR should not be included in a rollup?

**Comment by Centril** [comment]
@pnkfelix No you did it right, I just didn't look closely enough.

It would be good if homu enforced this automatically and refused to make the rollup... cc @kennytm

**Comment by kennytm** [comment]
@Centril Please file a feature request on rust-lang/homu

**Pull request merged**